compiler/rustc_typeck/src/lib.rs

```diff
@@ -164,191 +164,250 @@ fn require_same_types<'tcx>(
-    let main_t = tcx.type_of(main_def_id);
-    match main_t.kind() {
-        ty::FnDef(..) => {
-            if let Some(Node::Item(it)) = tcx.hir().find(main_id) {
-                if let hir::ItemKind::Fn(ref sig, ref generics, _) = it.kind {
-                    let mut error = false;
-                    if !generics.params.is_empty() {
-                        let msg = "`main` function is not allowed to have generic \
-                                   parameters"
-                            .to_owned();
-                        let label = "`main` cannot have generic parameters".to_string();
-                        struct_span_err!(tcx.sess, generics.span, E0131, "{}", msg)
-                            .span_label(generics.span, label)
-                            .emit();
-                        error = true;
-                    }
-                    if let Some(sp) = generics.where_clause.span() {
-                        struct_span_err!(
-                            tcx.sess,
-                            sp,
-                            E0646,
-                            "`main` function is not allowed to have a `where` clause"
-                        )
-                        .span_label(sp, "`main` cannot have a `where` clause")
-                        .emit();
-                        error = true;
-                    }
-                    if let hir::IsAsync::Async = sig.header.asyncness {
-                        let span = tcx.sess.source_map().guess_head_span(it.span);
-                        struct_span_err!(
-                            tcx.sess,
-                            span,
-                            E0752,
-                            "`main` function is not allowed to be `async`"
-                        )
-                        .span_label(span, "`main` function is not allowed to be `async`")
-                        .emit();
-                        error = true;
-                    }
-
-                    let attrs = tcx.hir().attrs(main_id);
-                    for attr in attrs {
-                        if tcx.sess.check_name(attr, sym::track_caller) {
-                            tcx.sess
-                                .struct_span_err(
-                                    attr.span,
-                                    "`main` function is not allowed to be `#[track_caller]`",
-                                )
-                                .span_label(
-                                    main_span,
-                                    "`main` function is not allowed to be `#[track_caller]`",
-                                )
-                                .emit();
-                            error = true;
-                        }
-                    }
-
-                    if error {
-                        return;
-                    }
-                }
+    let main_diagnostics_hir_id;
+    if let Some(main_local_def_id) = main_def_id.as_local() {
```

Comment:
I don't like that all the checking is done in two versions - one on HIR and one on `DefId` query results, it's very easy to get diverging implementations.
All the queries should work on local `DefId`s as well, so couldn't the implementations be merged into one and always work on queries without involving HIR?

Reply:
@petrochenkov I don't like this either... I might be wrong but it seems the function signature's original form (asyncness and where clauses) got transformed into other forms after the lowering and encoding into metadata. It seems impossible to report the exact accurate diagnostics for foreign mains. I'm not sure who i should ask to confirm this. maybe @estebank ?

Reply:
Could you make a function that encapsulates all the differences between local and extern cases and only returns diagnostic spans? I think the checking logic implementing user-visible language behavior should be unified.

Reply:
Done with the unification.

Reply:
My previously thinking was wrong... All the checks can be migrated to using queries to finish those checks. Now the logic is properly unified.